diff --git a/packages/common/pipes/index.ts b/packages/common/pipes/index.ts
index f416b9467..982807dfc 100644
--- a/packages/common/pipes/index.ts
+++ b/packages/common/pipes/index.ts
@@ -1,3 +1,4 @@
+export * from './abstract-parse.pipe';
 export * from './default-value.pipe';
 export * from './file';
 export * from './parse-array.pipe';
diff --git a/packages/common/pipes/parse-bool.pipe.ts b/packages/common/pipes/parse-bool.pipe.ts
index 177ee7cbe..3a52a062c 100644
--- a/packages/common/pipes/parse-bool.pipe.ts
+++ b/packages/common/pipes/parse-bool.pipe.ts
@@ -1,37 +1,15 @@
 import { Injectable } from '../decorators/core/injectable.decorator';
 import { Optional } from '../decorators/core/optional.decorator';
-import { HttpStatus } from '../enums/http-status.enum';
+import { ArgumentMetadata } from '../interfaces/features/pipe-transform.interface';
 import {
-  ArgumentMetadata,
-  PipeTransform,
-} from '../interfaces/features/pipe-transform.interface';
-import {
-  ErrorHttpStatusCode,
-  HttpErrorByCode,
-} from '../utils/http-error-by-code.util';
-import { isNil } from '../utils/shared.utils';
+  AbstractParsePipe,
+  AbstractParsePipeOptions,
+} from './abstract-parse.pipe';
 
 /**
  * @publicApi
  */
-export interface ParseBoolPipeOptions {
-  /**
-   * The HTTP status code to be used in the response when the validation fails.
-   */
-  errorHttpStatusCode?: ErrorHttpStatusCode;
-  /**
-   * A factory function that returns an exception object to be thrown
-   * if validation fails.
-   * @param error Error message
-   * @returns The exception object
-   */
-  exceptionFactory?: (error: string) => any;
-  /**
-   * If true, the pipe will return null or undefined if the value is not provided
-   * @default false
-   */
-  optional?: boolean;
-}
+export interface ParseBoolPipeOptions extends AbstractParsePipeOptions {}
 
 /**
  * Defines the built-in ParseBool Pipe
@@ -41,19 +19,12 @@ export interface ParseBoolPipeOptions {
  * @publicApi
  */
 @Injectable()
-export class ParseBoolPipe implements PipeTransform<
+export class ParseBoolPipe extends AbstractParsePipe<
   string | boolean,
-  Promise<boolean>
+  boolean
 > {
-  protected exceptionFactory: (error: string) => any;
-
   constructor(@Optional() protected readonly options?: ParseBoolPipeOptions) {
-    options = options || {};
-    const { exceptionFactory, errorHttpStatusCode = HttpStatus.BAD_REQUEST } =
-      options;
-    this.exceptionFactory =
-      exceptionFactory ||
-      (error => new HttpErrorByCode[errorHttpStatusCode](error));
+    super(options);
   }
 
   /**
@@ -66,8 +37,8 @@ export class ParseBoolPipe implements PipeTransform<
   async transform(
     value: string | boolean,
     metadata: ArgumentMetadata,
-  ): Promise<boolean> {
-    if (isNil(value) && this.options?.optional) {
+  ): Promise<boolean | string> {
+    if (this.isOptionalAndNil(value)) {
       return value;
     }
     if (this.isTrue(value)) {
diff --git a/packages/common/pipes/parse-date.pipe.ts b/packages/common/pipes/parse-date.pipe.ts
index 4dc1d9860..11894a59e 100644
--- a/packages/common/pipes/parse-date.pipe.ts
+++ b/packages/common/pipes/parse-date.pipe.ts
@@ -1,48 +1,25 @@
 import { Injectable } from '../decorators/core/injectable.decorator';
-import { HttpStatus } from '../enums/http-status.enum';
-import { PipeTransform } from '../interfaces/features/pipe-transform.interface';
-import {
-  ErrorHttpStatusCode,
-  HttpErrorByCode,
-} from '../utils/http-error-by-code.util';
+import { ArgumentMetadata } from '../interfaces/features/pipe-transform.interface';
 import { isNil } from '../utils/shared.utils';
+import {
+  AbstractParsePipe,
+  AbstractParsePipeOptions,
+} from './abstract-parse.pipe';
 
-export interface ParseDatePipeOptions {
-  /**
-   * If true, the pipe will return null or undefined if the value is not provided
-   * @default false
-   */
-  optional?: boolean;
+export interface ParseDatePipeOptions extends AbstractParsePipeOptions {
   /**
    * Default value for the date
    */
   default?: () => Date;
-  /**
-   * The HTTP status code to be used in the response when the validation fails.
-   */
-  errorHttpStatusCode?: ErrorHttpStatusCode;
-  /**
-   * A factory function that returns an exception object to be thrown
-   * if validation fails.
-   * @param error Error message
-   * @returns The exception object
-   */
-  exceptionFactory?: (error: string) => any;
 }
 
 @Injectable()
-export class ParseDatePipe implements PipeTransform<
-  string | number | undefined | null
+export class ParseDatePipe extends AbstractParsePipe<
+  string | number | undefined | null,
+  Date
 > {
-  protected exceptionFactory: (error: string) => any;
-
   constructor(private readonly options: ParseDatePipeOptions = {}) {
-    const { exceptionFactory, errorHttpStatusCode = HttpStatus.BAD_REQUEST } =
-      options;
-
-    this.exceptionFactory =
-      exceptionFactory ||
-      (error => new HttpErrorByCode[errorHttpStatusCode](error));
+    super(options);
   }
 
   /**
@@ -52,10 +29,11 @@ export class ParseDatePipe implements PipeTransform<
    * @param value currently processed route argument
    * @param metadata contains metadata about the currently processed route argument
    */
-  transform(
+  async transform(
     value: string | number | undefined | null,
-  ): Date | null | undefined {
-    if (this.options.optional && isNil(value)) {
+    metadata: ArgumentMetadata,
+  ): Promise<Date | null | undefined> {
+    if (this.isOptionalAndNil(value)) {
       return this.options.default ? this.options.default() : value;
     }
 
diff --git a/packages/common/pipes/parse-enum.pipe.ts b/packages/common/pipes/parse-enum.pipe.ts
index 47f3c263a..087808b1e 100644
--- a/packages/common/pipes/parse-enum.pipe.ts
+++ b/packages/common/pipes/parse-enum.pipe.ts
@@ -1,33 +1,15 @@
-import { Injectable, Optional } from '../decorators/core';
-import { ArgumentMetadata, HttpStatus } from '../index';
-import { PipeTransform } from '../interfaces/features/pipe-transform.interface';
+import { Injectable } from '../decorators/core/injectable.decorator';
+import { Optional } from '../decorators/core/optional.decorator';
+import { ArgumentMetadata } from '../interfaces/features/pipe-transform.interface';
 import {
-  ErrorHttpStatusCode,
-  HttpErrorByCode,
-} from '../utils/http-error-by-code.util';
-import { isNil } from '../utils/shared.utils';
+  AbstractParsePipe,
+  AbstractParsePipeOptions,
+} from './abstract-parse.pipe';
 
 /**
  * @publicApi
  */
-export interface ParseEnumPipeOptions {
-  /**
-   * If true, the pipe will return null or undefined if the value is not provided
-   * @default false
-   */
-  optional?: boolean;
-  /**
-   * The HTTP status code to be used in the response when the validation fails.
-   */
-  errorHttpStatusCode?: ErrorHttpStatusCode;
-  /**
-   * A factory function that returns an exception object to be thrown
-   * if validation fails.
-   * @param error Error message
-   * @returns The exception object
-   */
-  exceptionFactory?: (error: string) => any;
-}
+export interface ParseEnumPipeOptions extends AbstractParsePipeOptions {}
 
 /**
  * Defines the built-in ParseEnum Pipe
@@ -37,24 +19,17 @@ export interface ParseEnumPipeOptions {
  * @publicApi
  */
 @Injectable()
-export class ParseEnumPipe<T = any> implements PipeTransform<T> {
-  protected exceptionFactory: (error: string) => any;
+export class ParseEnumPipe<T = any> extends AbstractParsePipe<T, T> {
   constructor(
     protected readonly enumType: T,
     @Optional() protected readonly options?: ParseEnumPipeOptions,
   ) {
+    super(options);
     if (!enumType) {
       throw new Error(
         `"ParseEnumPipe" requires "enumType" argument specified (to validate input values).`,
       );
     }
-    options = options || {};
-    const { exceptionFactory, errorHttpStatusCode = HttpStatus.BAD_REQUEST } =
-      options;
-
-    this.exceptionFactory =
-      exceptionFactory ||
-      (error => new HttpErrorByCode[errorHttpStatusCode](error));
   }
 
   /**
@@ -65,7 +40,7 @@ export class ParseEnumPipe<T = any> implements PipeTransform<T> {
    * @param metadata contains metadata about the currently processed route argument
    */
   async transform(value: T, metadata: ArgumentMetadata): Promise<T> {
-    if (isNil(value) && this.options?.optional) {
+    if (this.isOptionalAndNil(value)) {
       return value;
     }
     if (!this.isEnum(value)) {
diff --git a/packages/common/pipes/parse-float.pipe.ts b/packages/common/pipes/parse-float.pipe.ts
index 04f5e97e2..2f4f5d3da 100644
--- a/packages/common/pipes/parse-float.pipe.ts
+++ b/packages/common/pipes/parse-float.pipe.ts
@@ -1,33 +1,16 @@
-import { Injectable, Optional } from '../decorators/core';
-import { ArgumentMetadata, HttpStatus } from '../index';
-import { PipeTransform } from '../interfaces/features/pipe-transform.interface';
+import { Injectable } from '../decorators/core/injectable.decorator';
+import { Optional } from '../decorators/core/optional.decorator';
+import { ArgumentMetadata } from '../interfaces/features/pipe-transform.interface';
+import { isFinite } from '../utils/shared.utils';
 import {
-  ErrorHttpStatusCode,
-  HttpErrorByCode,
-} from '../utils/http-error-by-code.util';
-import { isNil } from '../utils/shared.utils';
+  AbstractParsePipe,
+  AbstractParsePipeOptions,
+} from './abstract-parse.pipe';
 
 /**
  * @publicApi
  */
-export interface ParseFloatPipeOptions {
-  /**
-   * The HTTP status code to be used in the response when the validation fails.
-   */
-  errorHttpStatusCode?: ErrorHttpStatusCode;
-  /**
-   * A factory function that returns an exception object to be thrown
-   * if validation fails.
-   * @param error Error message
-   * @returns The exception object
-   */
-  exceptionFactory?: (error: string) => any;
-  /**
-   * If true, the pipe will return null or undefined if the value is not provided
-   * @default false
-   */
-  optional?: boolean;
-}
+export interface ParseFloatPipeOptions extends AbstractParsePipeOptions {}
 
 /**
  * Defines the built-in ParseFloat Pipe
@@ -37,17 +20,9 @@ export interface ParseFloatPipeOptions {
  * @publicApi
  */
 @Injectable()
-export class ParseFloatPipe implements PipeTransform<string> {
-  protected exceptionFactory: (error: string) => any;
-
+export class ParseFloatPipe extends AbstractParsePipe<string, number> {
   constructor(@Optional() protected readonly options?: ParseFloatPipeOptions) {
-    options = options || {};
-    const { exceptionFactory, errorHttpStatusCode = HttpStatus.BAD_REQUEST } =
-      options;
-
-    this.exceptionFactory =
-      exceptionFactory ||
-      (error => new HttpErrorByCode[errorHttpStatusCode](error));
+    super(options);
   }
 
   /**
@@ -57,8 +32,8 @@ export class ParseFloatPipe implements PipeTransform<string> {
    * @param value currently processed route argument
    * @param metadata contains metadata about the currently processed route argument
    */
-  async transform(value: string, metadata: ArgumentMetadata): Promise<number> {
-    if (isNil(value) && this.options?.optional) {
+  async transform(value: string, metadata: ArgumentMetadata): Promise<number | string> {
+    if (this.isOptionalAndNil(value)) {
       return value;
     }
     if (!this.isNumeric(value)) {
diff --git a/packages/common/pipes/parse-int.pipe.ts b/packages/common/pipes/parse-int.pipe.ts
index d74aa0205..2a265970b 100644
--- a/packages/common/pipes/parse-int.pipe.ts
+++ b/packages/common/pipes/parse-int.pipe.ts
@@ -1,37 +1,16 @@
 import { Injectable } from '../decorators/core/injectable.decorator';
 import { Optional } from '../decorators/core/optional.decorator';
-import { HttpStatus } from '../enums/http-status.enum';
+import { ArgumentMetadata } from '../interfaces/features/pipe-transform.interface';
+import { isFinite } from '../utils/shared.utils';
 import {
-  ArgumentMetadata,
-  PipeTransform,
-} from '../interfaces/features/pipe-transform.interface';
-import {
-  ErrorHttpStatusCode,
-  HttpErrorByCode,
-} from '../utils/http-error-by-code.util';
-import { isNil } from '../utils/shared.utils';
+  AbstractParsePipe,
+  AbstractParsePipeOptions,
+} from './abstract-parse.pipe';
 
 /**
  * @publicApi
  */
-export interface ParseIntPipeOptions {
-  /**
-   * The HTTP status code to be used in the response when the validation fails.
-   */
-  errorHttpStatusCode?: ErrorHttpStatusCode;
-  /**
-   * A factory function that returns an exception object to be thrown
-   * if validation fails.
-   * @param error Error message
-   * @returns The exception object
-   */
-  exceptionFactory?: (error: string) => any;
-  /**
-   * If true, the pipe will return null or undefined if the value is not provided
-   * @default false
-   */
-  optional?: boolean;
-}
+export interface ParseIntPipeOptions extends AbstractParsePipeOptions {}
 
 /**
  * Defines the built-in ParseInt Pipe
@@ -41,17 +20,9 @@ export interface ParseIntPipeOptions {
  * @publicApi
  */
 @Injectable()
-export class ParseIntPipe implements PipeTransform<string> {
-  protected exceptionFactory: (error: string) => any;
-
+export class ParseIntPipe extends AbstractParsePipe<string, number> {
   constructor(@Optional() protected readonly options?: ParseIntPipeOptions) {
-    options = options || {};
-    const { exceptionFactory, errorHttpStatusCode = HttpStatus.BAD_REQUEST } =
-      options;
-
-    this.exceptionFactory =
-      exceptionFactory ||
-      (error => new HttpErrorByCode[errorHttpStatusCode](error));
+    super(options);
   }
 
   /**
@@ -61,8 +32,8 @@ export class ParseIntPipe implements PipeTransform<string> {
    * @param value currently processed route argument
    * @param metadata contains metadata about the currently processed route argument
    */
-  async transform(value: string, metadata: ArgumentMetadata): Promise<number> {
-    if (isNil(value) && this.options?.optional) {
+  async transform(value: string, metadata: ArgumentMetadata): Promise<number | string> {
+    if (this.isOptionalAndNil(value)) {
       return value;
     }
     if (!this.isNumeric(value)) {
diff --git a/packages/common/pipes/parse-uuid.pipe.ts b/packages/common/pipes/parse-uuid.pipe.ts
index 1c67860b9..db4a0d631 100644
--- a/packages/common/pipes/parse-uuid.pipe.ts
+++ b/packages/common/pipes/parse-uuid.pipe.ts
@@ -1,40 +1,20 @@
 import { Injectable } from '../decorators/core/injectable.decorator';
 import { Optional } from '../decorators/core/optional.decorator';
-import { HttpStatus } from '../enums/http-status.enum';
+import { ArgumentMetadata } from '../interfaces/features/pipe-transform.interface';
+import { isString } from '../utils/shared.utils';
 import {
-  ArgumentMetadata,
-  PipeTransform,
-} from '../interfaces/features/pipe-transform.interface';
-import {
-  ErrorHttpStatusCode,
-  HttpErrorByCode,
-} from '../utils/http-error-by-code.util';
-import { isNil, isString } from '../utils/shared.utils';
+  AbstractParsePipe,
+  AbstractParsePipeOptions,
+} from './abstract-parse.pipe';
 
 /**
  * @publicApi
  */
-export interface ParseUUIDPipeOptions {
+export interface ParseUUIDPipeOptions extends AbstractParsePipeOptions {
   /**
    * UUID version to validate
    */
   version?: '3' | '4' | '5' | '7';
-  /**
-   * The HTTP status code to be used in the response when the validation fails.
-   */
-  errorHttpStatusCode?: ErrorHttpStatusCode;
-  /**
-   * A factory function that returns an exception object to be thrown
-   * if validation fails.
-   * @param error Error message
-   * @returns The exception object
-   */
-  exceptionFactory?: (errors: string) => any;
-  /**
-   * If true, the pipe will return null or undefined if the value is not provided
-   * @default false
-   */
-  optional?: boolean;
 }
 
 /**
@@ -45,7 +25,7 @@ export interface ParseUUIDPipeOptions {
  * @publicApi
  */
 @Injectable()
-export class ParseUUIDPipe implements PipeTransform<string> {
+export class ParseUUIDPipe extends AbstractParsePipe<string, string> {
   protected static uuidRegExps = {
     3: /^[0-9A-F]{8}-[0-9A-F]{4}-3[0-9A-F]{3}-[0-9A-F]{4}-[0-9A-F]{12}$/i,
     4: /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i,
@@ -54,24 +34,14 @@ export class ParseUUIDPipe implements PipeTransform<string> {
     all: /^[0-9A-F]{8}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{12}$/i,
   };
   private readonly version: '3' | '4' | '5' | '7' | undefined;
-  protected exceptionFactory: (errors: string) => any;
 
   constructor(@Optional() protected readonly options?: ParseUUIDPipeOptions) {
-    options = options || {};
-    const {
-      exceptionFactory,
-      errorHttpStatusCode = HttpStatus.BAD_REQUEST,
-      version,
-    } = options;
-
-    this.version = version;
-    this.exceptionFactory =
-      exceptionFactory ||
-      (error => new HttpErrorByCode[errorHttpStatusCode](error));
+    super(options);
+    this.version = options?.version;
   }
 
   async transform(value: string, metadata: ArgumentMetadata): Promise<string> {
-    if (isNil(value) && this.options?.optional) {
+    if (this.isOptionalAndNil(value)) {
       return value;
     }
     if (!this.isUUID(value, this.version)) {
