diff --git a/packages/websockets/index.ts b/packages/websockets/index.ts
index c831a18b2..4569b5df9 100644
--- a/packages/websockets/index.ts
+++ b/packages/websockets/index.ts
@@ -10,5 +10,6 @@ export * from './adapters';
 export * from './decorators';
 export * from './errors';
 export * from './exceptions';
+export * from './interceptors';
 export { MessageMappingProperties } from './gateway-metadata-explorer';
 export * from './interfaces';
